Firefox Support

mlapaglia opened the pull request:
# 🚀 Enhanced Build Process, Automated Releases, and Firefox Support

## Overview
This PR introduces a comprehensive overhaul of the build and release process, along with enhanced Firefox support for the Stream Deck Google Meet plugin. The changes eliminate manual version management, introduce conditional store publishing, and streamline the CI/CD pipeline.

## ✨ Key Features

### 🏷️ **Dynamic Version Management**
- **Automatic version extraction** from git tags (e.g., `v1.7.0` → `1.7.0`)
- **No more hardcoded versions** in `manifest.json` or `package.json`
- **Version injection** into manifest files during build
- **Fallback handling** to package.json version when no git tags are available

### 🔄 **Automated Release Process**
- **Conditional publishing** based on GitHub release/prerelease flag
- **Pre-releases** while pull requests are active
- **Automatic artifact attachment** to GitHub releases

### 🦊 **Added Firefox Support**
- **Added Firefox extension building** and packaging
- **Firefox-specific manifest handling** (`manifest_firefox.json`)
- **Web-ext integration** for linting and validation, and signing

## 🎯 **Benefits**

### **For Developers**
- **Zero manual version management** - versions come from git tags
- **Faster CI/CD feedback** - parallel builds and comprehensive validation
- **Clear artifact downloads** - test extensions directly from PR build

### **For Releases**
- **Mistake-proof publishing** - prerelease flag controls store publishing
- **Version consistency** - same version across all components
- **Comprehensive testing** - all components validated before release

## 🔄 **Migration Guide**

### **Old Process**
```bash
# Manual version bumping
# Edit manifest.json, package.json
# Commit changes
# Manual workflow dispatch for publishing
```

### **New Process**
```bash
# Create release with semantic version tag
git tag v1.7.0
git push origin v1.7.0

# Create GitHub release
# ✅ Prerelease unchecked → Auto-publish to stores
# ☑️ Prerelease checked → Manual download only
```

## 🧪 **Testing Done**
- ✅ Version extraction from git tags
- ✅ Build process with dynamic versioning
- ✅ Manifest validation for both Chrome and Firefox
- ✅ Extension packaging (ZIP creation)
- ✅ Artifact upload and retention
- ✅ Firefox linting and signing integration
- ✅ Multi-platform plugin builds

## 🚀 **Ready for Review**
This PR is ready for review and testing. The new workflows have been tested and provide a robust foundation for ongoing development and releases.

mlapaglia commented:
You'll need to do a little work getting the extensions set up the first time in their respective stores, but after that deployments should be automated when releases  are created!

ChrisRegado commented on .github/workflows/streamdeck-plugin-build.yml:
```diff
@@ -73,3 +118,55 @@ jobs:
         with:
           name: com.chrisregado.googlemeet.streamDeckPlugin
           path: output/com.chrisregado.googlemeet.streamDeckPlugin
+      - name: Upload browser extensions
```
Could this be moved to the end of `build-browser-extensions` job instead of downloading & re-uploading in `bundle-streamdeck-distributable` ? `bundle-streamdeck-distributable` is all about running the Elgato packaging tool which otherwise hasn't changed in this PR. (It's unrelated to browser extension management.)

ChrisRegado commented on .github/workflows/browser-extension-release.yml:
```diff
@@ -0,0 +1,135 @@
+          publish: true
+          publish-target: default
+      
+      - name: Publish to Firefox Add-ons
```
If we're not going to be publishing to the store and we already use `web-ext`, would it work to replace this with a call to `web-ext sign --channel=unlisted` as part of the build job? (I'm not super familiar with web-ext, but I was looking at [these docs](https://extensionworkshop.com/documentation/develop/getting-started-with-web-ext/#sign-for-self-distribution).)

ChrisRegado commented:
Thanks for posting!

> Do you want to automatically publish the plugin to the store also instead of installing it manually?

I'll get AMO signing set up for Firefox .xpi [self-distribution](https://extensionworkshop.com/documentation/publish/self-distribution/) (which the docs suggest will let installations persist across restarts), but I don't want to publicly list the extensions on the Google/Mozilla stores. That means we can remove the Chrome extension publishing.

I haven't been using Google Meet or even my Stream Deck for almost 3 years now outside of testing this plugin for (rare) pull requests or addressing reports of major breakage. To be honest, I was already considering archiving the GitHub project... So I don't want the management burden or implication of support of having public store listings at this time. This plugin is effectively in minimal maintenance mode.

ChrisRegado commented:
It's mostly just that this is a small free side project and I have minimal time to spend on it.

Posting it on the stores brings with it:

* More support questions or complaints in the form of bad reviews every time Google updates the Meet UI and the extension breaks, or from users not understanding that you need both the browser extension and the Stream Deck plugin for it to work, etc etc.
* Likely issues with using Google trademarks and icons, especially on the Chrome store. (https://developer.chrome.com/docs/webstore/branding/#choosing-an-extension-or-developer-name) We'll probably need to rebrand it and maybe replace all the icons in the plugin.

(See also https://github.com/ChrisRegado/streamdeck-googlemeet/issues/13.)

Sharing it "as-is" on GitHub definitely limits usage down to a smaller more technically inclined audience, but I initially wrote it only for myself to use, so I figured uploading it here was still a net positive. :) 

If you and/or others are able to commit to longer-term maintenance of the project, then publishing to the stores (including the Elgato store) becomes more viable. I could eventually migrate the repo to its own GitHub Organization or something like that. So far, no repeat involvement in the project has materialized.

mlapaglia commented:
alright, i'll simplify the pipeline down then for now. so for now just make the pipeline publish the plugin, the chrome installation process will continue as is, and you'll manually update the release with the firefox extension once you get it signed?

ChrisRegado commented:
> the pipeline publish the plugin, the chrome installation process will continue as is

Yup. Most of this PR still works -- having a zip of just the (unsigned) Chrome extension to attach to a release is a nice improvement, even if users still have to do the `Load Unpacked` approach to install it.

> you'll manually update the release with the firefox extension once you get it signed?

I'm happy to still have a CI job to do the package signing. I added `FIREFOX_ADDON_GUID`, `FIREFOX_JWT_ISSUER`, and `FIREFOX_JWT_SECRET` secrets to the GitHub repo. Heads-up, I changed `manifest_firefox.json` to set `"id": "{233c4f86-64fb-11f0-ad1e-00155d1f8e7e}",` for initial submission to Mozilla to make it more generic in case we end up moving the repo to a GitHub org.

The requested change is just making it unlisted.

---

For reference, I tested the following locally (with the updated gecko ID) to validate we can still do automated signing without having a public store listing:

```
export WEB_EXT_API_KEY="<redacted>"       # set to secrets.FIREFOX_JWT_ISSUER
export WEB_EXT_API_SECRET="<redacted>"    # set to secrets.FIREFOX_JWT_SECRET
npm run build
web-ext sign --channel=unlisted --source-dir=build/firefox --artifacts-dir=build
```

Output:
```
Building web extension from build/firefox
(node:16653) ExperimentalWarning: buffer.File is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Waiting for validation...
Waiting for approval...
Signed xpi downloaded: build/streamdeck_googlemeet-1.6.0.xpi
```

The `sign` command takes a minute or two, but the result is a local signed .xpi file that's ready to attach to the Release. I validated it's installable on the vanilla Firefox release version with no config changes and persists across browser restarts.

mlapaglia commented:
ok i've reverted most of it. now when a release is created and pointed at the branch it will create a prerelease [like this](https://github.com/mlapaglia/streamdeck-googlemeet/releases/tag/v0.0.7) and the workflow will trigger, build the extensions, and upload them to the release.

how do we want to handle the plugin? should we build the extensions and plugin every time regardless if they changed or not?

ChrisRegado commented on .github/workflows/browser-extension-release.yml:
```diff
@@ -0,0 +1,80 @@
+        uses: actions/upload-artifact@v4
+        with:
+          name: firefox-extension-${{ steps.get-version.outputs.version }}
+          path: browser-extension/build/firefox-extension.zip
```
For releases, we want to attach the signed `.xpi` file from `npm run sign-firefox` command above, right? So something like `browser-extension/build/streamdeck_googlemeet-${{ steps.get-version.outputs.version }}.xpi` or the `mv build/streamdeck_googlemeet*.xpi build/firefox-extension.xpi` trick.

ChrisRegado reviewed (commented):
Nice, thanks!

> how do we want to handle the plugin? should we build the extensions and plugin every time regardless if they changed or not? whenever a user opens up the release they'll have everything they need available as an artifact in there.

Yeah we can bundle it every time.

Little tip: The `bundle-streamdeck-distributable` job from `streamdeck-plugin-build.yml` currently doesn't use any secrets (Elgato doesn't have any signing), so it's safe to do the same sdPlugin bundling procedure both for PRs and for the release job.

Also, feel free to handle the plugin in a different PR if you want to get the browser extension stuff wrapped up & merged. Up to you!

mlapaglia commented:
i think i understand the direction you want me to take, i am a novice GHA developer 😆! i fixed the file extensions and removed  manually settings the secrets.
